### ISSUE TITLE
NuGet for building starcounter

### ISSUE THREAD

**Issue by careri** [Starcounter/Starcounter#3812]
Hi!

I've made some scripts that extracts the needed file from a Starcounter(2) installation and creates a NuGet using the VersionInfo.xml to define the version of the NuGet.

The reason for doing this is that we want our build-server (Jenkins) to output the assemblies that we distribute to our customers.

But I don't want to install Starcounter on the buildserver. The other reason for using a NuGet is to support different versions of Starcounter in different Git-branches.

I added a Props file that checks if the env-var StarcounterBin is defined, if so I really on the assemblies at that location. If the system is missing the StarcounterBin var I fallback to the assemblies in my NuGet.

I got the scripts to make them easy enough, but it would be even easier if you had a NuGet channel for distributing this kind of NuGet.

This is the content of my NuGet:
>
packages\Heads.Scdb2.Build.2.2.1834\tools\FasterThanJson.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\FSharp.Core.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Mono.CSharp.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\scerrres.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.BitsAndBytes.Unsafe.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.Hosting.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.Internal.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.MsBuild.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.MsBuild.targets
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.REST.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.XSON.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.XSON.JsonByExample.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.XSON.JsTemplateParser.dll
packages\Heads.Scdb2.Build.2.2.1834\tools\Starcounter.XSON.PartialClassGenerator.dll
packages\Heads.Scdb2.Build.2.2.1834\build\Heads.Scdb2.Build.Props

**Comment by per-samuelsson** [comment]
Hi @careri,

I'm not sure I understand what you request here, nor what you are trying to do? Are you asking if it would be possible to develop apps on Starcounter without doing an installation, just relying on a NuGet package (or packages)?

If so, we have certainly had discussions in such direction and would like a more lightweight experience long term. But we don't have anything that support that right now really.

As for the list, it seem incomplete, even not considering that it lack all tooling (to create and manage databases). As an example, I don't see any of our native binaries in it.

**Comment by careri** [comment]
Nah, it's build server related. 
We develop on machines with Starcounter installed. 

A real case for us now is that we
re going to use 2.3 for Retail. While I haven't migrated my product from 2.2 yet.

The build server must be able to build both products. This could be done with two installations of starcounter and removing the StarcounterBin environement var and instead pass it on build command line.

But my NuGet solution is cleaner and easier to use.
I just bind Heads.Scdb.Build.2.2.1834 to my product and let's say Heads.Scdb.Build.2.3.11 to the Retail product.

I don't need to install starcounter, I just need the references during build. The references aren't copied to the output of the build server.

I still need a machine with a valid starcounter installation to run the output from the build server.

**Comment by per-samuelsson** [comment]
Aha, ok, thats clearly a different case, but now I understand it fully. It's more marginal than that of supporting the model I proposed above.

Let us consider it a little and get back to you.

**Comment by per-samuelsson** [comment]
The case you describe actually share some ideas we've discussed allowing Starcounter applications to be developed using just abstractions, and that you have to take a certain action to run it in a real installation - and that installation could very possibly be hosted in the cloud.

That's a good thing, since it make it more likely that this story is something we want to eventually address.

**Comment by careri** [comment]
Yes, using interfaces/abstractions for the development would work nicely.
If you put such interfaces in a NuGet your server code could depend on that NuGet. And the same NuGet could be distributed to developers that are not running the server locally.
It would also solve the build server case discussed here.
